src/libsystemd-network/dhcp-option.c

```diff
@@ -356,10 +349,167 @@ static int parse_options(const uint8_t options[], size_t buflen, uint8_t *overlo
         return 0;
 }
 
+struct MergeOptionNode {
+        const uint8_t *data;
+        LIST_FIELDS(struct MergeOptionNode, options);
+};
+
+static struct MergeOptionNode* merge_option_free(struct MergeOptionNode *head) {
+        LIST_CLEAR(options, head, free);
+        return NULL;
+}
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(merge_option_node_ops,
+                                              void,
+                                              trivial_hash_func,
+                                              trivial_compare_func,
+                                              struct MergeOptionNode,
+                                              merge_option_free);
+
+static int dhcp_option_merge_append(const uint8_t *buf, size_t buflen, OrderedHashmap *opts,
+                                    size_t *tot_len, uint8_t *overload) {
+        size_t offset = 0;
+        struct MergeOptionNode *head, *next;
+
+        while (offset < buflen) {
+                const uint8_t *data = &buf[offset];
+                uint8_t code = buf[offset++];
+
+                switch (code) {
+                case SD_DHCP_OPTION_PAD:
+                        continue;
+                case SD_DHCP_OPTION_END:
+                        return 0;
+                }
+
+                if (buflen < offset + 1)
+                        return -ENOBUFS;
+
+                uint8_t len = buf[offset++];
+
+                if (buflen < offset + len)
+                        return -EINVAL;
+
+                switch (code) {
+                case SD_DHCP_OPTION_OVERLOAD:
+                        if (len != 1)
+                                return -EINVAL;
+                        if (overload)
+                                *overload = buf[offset];
+                        offset++;
+                        continue;
+                }
+
+                next = new(struct MergeOptionNode, 1);
+                if (!next)
+                        return -ENOMEM;
+                LIST_INIT(options, next);
+                next->data = data;
```

Comment:
maybe use compound init to initialize the struct. no need to initialize the list explicitly then

Reply:
Done.